gradio/components.py

```diff
@@ -2500,21 +2500,21 @@ def update(
     def preprocess(self, x: List[List[str | Number | bool]]):
         """
         Parameters:
-        x (List[List[str | number | bool]]): 2D array of str, numeric, or bool data
+        x (Dict[headers: List[str], data: (List[List[str | number | bool]]]): 2D array of str, numeric, or bool data
         Returns:
         (pandas.DataFrame | numpy.array | List[str | float | bool], List[List[str | float | bool]]): Dataframe in requested format
         """
         if self.type == "pandas":
-            if self.headers:
-                return pd.DataFrame(x, columns=self.headers)
+            if x["headers"]:
+                return pd.DataFrame(x["data"], columns=x["headers"])
             else:
-                return pd.DataFrame(x)
+                return pd.DataFrame(x["data"])
         if self.col_count[0] == 1:
```

Comment:
I don't think it's correct to use the `self.col_count` defined when the component is created. If the headers are dynamic so is the `col_count`? 

In this repro, the function is only given access to the first column which is not the expected behavior imo

```python
import gradio as gr
import pandas as pd
import numpy as np

def return_empty(x):
    return np.atleast_1d(x.astype(float).mean(axis=0))


gr.Interface(return_empty,
             gr.Dataframe(type="numpy", headers=["a"]),
             gr.Dataframe(headers=["mean"])).launch()
```

Reply:
What was this logic ever even doing? This feels wrong regardless of the changes in this PR. Can we remove? @abidlabs @aliabid94